# sigsim

*sigsim* (aka *ssssim*) is a Stupidly Simple Signal Simulator for simulating nanopore raw signal data. It is currently in early development stage and therefore anticipate bugs and interface changes.

*sigsim* uses traditional pore models and gaussian noise for simulation. Due to simplicity, simulation would not be perfect, but takes miniscule effort to setup and run. You may checkout sophisticated deep-learning-based simulators such as [Nanosim](https://github.com/bcgsc/NanoSim) and [DeepSimulator](https://github.com/liyu95/DeepSimulator), if you wish.

Currently, reads directly extracted from the reference genome are simulated without any mutations/variants. If you want to have variants in your simulated data, you can first apply a set of variants to the reference using [bcftools](http://www.htslib.org/download/) and use that as an input to *sigsim*.

*sigsim* at the moment is single threaded and not optimised for efficient runtime performance or RAM usage. Generating 4000 reads from human genome takes like 2-3 minutes with around 3 GB of RAM.

## Building

```
sudo apt-get install zlib1g-dev   #install zlib development libraries
git clone https://github.com/hasindu2008/sigsim
cd sigsim
make
```

The commands to install zlib __development libraries__ on some popular distributions :
```sh
On Debian/Ubuntu : sudo apt-get install zlib1g-dev
On Fedora/CentOS : sudo dnf/yum install zlib-devel
On OS X : brew install zlib
```

## Usage

The simplest command to generate reads:
```
sigsim [OPTIONS] ref_genome.fa -o out_signal.blow5 -n NUM_READS
```

You can basecall the generated raw signal directly from the BLOW format using our fork of [dorado basecaller](https://github.com/hiruna72/dorado/releases/tag/v0.0.1) or using the SLOW5 guppy wrapper called [buttery-eel](https://github.com/Psy-Fer/buttery-eel).  Alternatively, if you love FAST5 that much, use [slow5tools](https://github.com/hasindu2008/slow5tools) to convert the BLOW5 to FAST5 and then use Guppy basecaller.

Generated read IDs encodes the true mapping positions in a format like `S1_33!chr1!225258409!225267761!-` which is the format compatible with [*mapeval* command in *paftools.js* under Minimap2 repository](https://github.com/lh3/minimap2/blob/master/misc/README.md#evaluation).

Basic options are as below:
- `-o FILE`: SLOW5/BLOW5 file to write
- `-x STR`: parameter profile (always applied before other options). e.g., dna-r9-min, dna-r9-prom, rna-r9-min, rna-r9-prom.
- `-n INT`: number of reads to simulate
- `-q FILE`: if you want to save the original reads directly taken from the reference genome (without basecalling errors) in FASTA format.
- `-r` the median read length (estimated median only, unused for direct RNA).
- `--ideal`: to generate perfect signals with no noise.


Advanced options are as below:
- `--full-contigs`: generate a complete raw signal per each contig in the input reference genome (incompatible with `-n`, only effective for DNA).
-  `--prefix`: generate prefixes such as adaptor (and polya for RNA)
-  `--kmer-model FILE`: custom nucleotide k-mer model file (format similar to https://github.com/hasindu2008/f5c/blob/master/test/r9-models/r9.4_450bps.nucleotide.6mer.template.model)
-  `--seed INT`: Seed or random generators (if 0, will be autogenerated)
-  `--ideal-time`: Generate signals with no time domain noise
-  `--ideal-amp `: Generate signals with no amplitude domain noise
-  `--dwell-mean FLOAT`: Mean number of signal samples per base

## Examples

DNA examples:

```
# generate 150,000 PromethION DNA reads from a reference genome
sigsim hg38noAlt.fa -x dna-r9-prom -o reads.blow5 -n 150000

# generate 30,000 MinION ultra-long DNA reads with median readlength of around 50,000 bases
sigsim hg38noAlt.fa -x dna-r9-min -o reads.blow5 -n 30000 -r 50000

# generate 1000 PromethION DNA reads with perfect signals with no noise
sigsim hg38noAlt.fa -x dna-r9-prom -o reads.blow5 -n 1000 --ideal

```

RNA examples:
```
# generate 4000 PromethION direct RNA reads from a transcriptome while including the adaptor and polyA tail
sigsim gencode.v40.transcripts.fa -x rna-r9-prom -o reads.blow5 -n 4000 --prefix
```

DNA example with variants that requires bcftools:

```
bcftools consensus -f hg38noAlt.fa na12878.vcf -o na12878.fa
sigsim na12878.fa -x dna-r9-prom -o reads.blow5 -n 150000
```

## Acknowledgement

The pore-models are from [Nanopolish](https://github.com/jts/nanopolish).
Some code snippets have been taken from [Minimap2](https://github.com/lh3/minimap2), [Samtools](http://samtools.sourceforge.net/).
Kseq from [klib](https://github.com/attractivechaos/klib) is used.